File: Kalk/Logica/spezzata.cpp
#include "spezzata.h"
#include <typeinfo>

Spezzata::Spezzata(const Punto& inizio,const Punto& fine):Linea(inizio,fine),punti(){
   // setTag(inizio.getTag()+fine.getTag());
}

bool Spezzata::trovaPunto(QString s)const{
    for(auto cit=punti.constBegin();cit!=punti.constEnd();++cit){
        if(cit->getTag()==s) return true;
    }
    return false;
}

Tag* Spezzata::operator+(const Tag& t) const{
    if(dynamic_cast<const Punto*>(&(t))){
        Spezzata* s=new Spezzata(*this);

        s->aggiungiPunto(s->getFine());
        s->setFine(static_cast<const Punto&>(t));
        return s;
    }
    if(typeid(Linea)==typeid(t)){
        Spezzata* s=new Spezzata(*this);
        s->aggiungiPunto(getFine());

        s->aggiungiPunto(static_cast<const Linea&>(t).getInizio());
        s->setFine(static_cast<const Linea&>(t).getFine());
        return s;
    }
    if(typeid(Spezzata)==typeid(t)){
        Spezzata* s=new Spezzata(*this);
        s->aggiungiPunto(getFine());
        s->aggiungiPunto(static_cast<const Linea&>(t).getInizio());
        for(auto cit=static_cast<const Spezzata&>(t).punti.constBegin();cit!=static_cast<const Spezzata&>(t).punti.constEnd();++cit){
            s->aggiungiPunto(*cit);
        }
        s->setFine(static_cast<const Linea&>(t).getFine());
        return s;
    }
}

QString Spezzata::output()const{  
    QString stringaTag,stringaCoord;

    stringaCoord=stringaCoord+" ("+QString::number(getInizio().x())+","+
            QString::number(getInizio().y())+"),";
    for(auto cit=punti.constBegin();cit!=punti.constEnd();++cit){
        stringaCoord=stringaCoord+"("+QString::number(cit->x())+","+
                QString::number(cit->y())+"),";
    }
    stringaCoord=stringaCoord+"("+QString::number(getFine().x())+","+QString::number(getFine().y())+")";
    return getTag()+stringaCoord;
}

Spezzata* Spezzata::simmetricoX(QString etichetta)const{
    Spezzata* s=new Spezzata(*(getInizio().simmetricoX(etichetta)),*(getFine().simmetricoX("")));
    for(auto cit=punti.constBegin();cit!=punti.constEnd();++cit){
        s->aggiungiPunto(*(cit->simmetricoX("")));
    }
    return s;
}
Spezzata* Spezzata::simmetricoY(QString etichetta)const{
    Spezzata* s=new Spezzata(*(getInizio().simmetricoY(etichetta)),*(getFine().simmetricoY("")));
    for(auto cit=punti.constBegin();cit!=punti.constEnd();++cit){
        s->aggiungiPunto(*(cit->simmetricoY("")));
    }
    return s;
}
Spezzata* Spezzata::simmetricoO(QString etichetta)const{
    Spezzata* s=new Spezzata(*(getInizio().simmetricoO(etichetta)),*(getFine().simmetricoO("")));
    for(auto cit=punti.constBegin();cit!=punti.constEnd();++cit){
        s->aggiungiPunto(*(cit->simmetricoO("")));
    }
    return s;
}
void Spezzata::traslazione(double nX,double nY){
    const_cast<Punto&>(getInizio()).traslazione(nX,nY);
    const_cast<Punto&>(getFine()).traslazione(nX,nY);
    for(auto it=punti.begin();it!=punti.end();++it){
        it->traslazione(nX,nY);
    }
}
double Spezzata::lunghezza()const{
    double somma=0;
    if(punti.length()==0) return getInizio().distanzaP(getFine());
    if(punti.length()==1) return getInizio().distanzaP(punti.first()) + punti.first().distanzaP(getFine());
    else{
        somma=somma+getInizio().distanzaP(punti.first());
        for(auto cit=punti.constBegin();cit!=punti.constEnd()-1;++cit){
            somma=somma+cit->distanzaP(*(cit+1));
        }
        somma=somma+getFine().distanzaP(punti.last());
        return somma;
    }
}

void Spezzata::aggiungiPunto(const Punto& p){
    punti.push_back(p);
}

Spezzata* Spezzata::chiudiSpezzata(QString etichetta) const{
    Spezzata* s=new Spezzata(getInizio(),getInizio());
    for(auto cit=punti.constBegin();cit!=punti.constEnd();++cit){
        s->aggiungiPunto(*(cit));
    }
    s->aggiungiPunto(getFine());
    s->setTag(etichetta);
    return s;
}

File: java/Kalk/src/Kalk/Classes/Spezzata.java
package Kalk.Classes;
import java.util.ArrayList;
import java.util.ListIterator;

public class Spezzata extends Linea {
	private ArrayList<Punto> punti=new ArrayList<Punto>();
	
	public Spezzata(String s, Punto i, Punto f) {
		super(s, i, f);
	}
	
	public ArrayList<Punto> getLista(){
		return punti;
	}
	

	public String toString() {
		String c="";
		c=c+" ("+getInizio().x()+","+getInizio().y()+"),";
		ListIterator<Punto> i;
		i=punti.listIterator();
		while(i.hasNext()) {
			c=c+"("+i.next().x();
			i.previous();
			c=c+","+i.next().y()+"),";
		}
		c=c+"("+getFine().x()+","+getFine().y()+")";
		return getTag()+c;
	}
	
	public void aggiungi(Punto p) {
		punti.add(p);
	}
	
	public Spezzata plus(Tag t) {
		if(t instanceof Punto) {
			aggiungi(getFine());
			setFine((Punto) t);
			return this;
		}
		if(t instanceof Spezzata) {
			aggiungi(getFine());
			aggiungi(((Linea)t).getInizio());
			((Spezzata)t).punti.forEach(Punto->{aggiungi(Punto);});
			setFine(((Linea)t).getFine());
			return this;
		}
		if(t instanceof Linea) {
			aggiungi(getFine());
			aggiungi(((Linea)t).getInizio());
			setFine(((Linea)t).getFine());
			return this;
		}
		return null;
	}
	
	@Override
	public Spezzata simmetricoX(String s) {
		Spezzata ns=new Spezzata(s,getInizio().simmetricoX(""),getFine().simmetricoX(""));
		punti.forEach((Punto)->{ns.aggiungi(Punto.simmetricoX(s));});
		return ns;
	}

	@Override
	public Spezzata simmetricoY(String s) {
		Spezzata ns=new Spezzata(s,getInizio().simmetricoY(""),getFine().simmetricoY(""));
		punti.forEach((Punto)->{ns.aggiungi(Punto.simmetricoY(s));});
		return ns;
	}

	@Override
	public Spezzata simmetricoO(String s) {
		Spezzata ns=new Spezzata(s,getInizio().simmetricoO(""),getFine().simmetricoO(""));
		punti.forEach((Punto)->{ns.aggiungi(Punto.simmetricoO(s));});
		return ns;
	}

	@Override
	public void traslazione(double nX, double nY) {
		getInizio().traslazione(nX, nY);
		getFine().traslazione(nX, nY);
		punti.forEach((Punto)->{Punto.traslazione(nX, nY);});
	}
	
	public double lunghezza(){
	    double somma=0;
	    if(punti.size()==0) return getInizio().distanzaP(getFine());
	    if(punti.size()==1) return getInizio().distanzaP(punti.get(0)) + punti.get(0).distanzaP(getFine());
	    else{
	        somma=somma+getInizio().distanzaP(punti.get(0));
	        for(int j=0; j!=punti.size()-2; j++){
	            somma=somma+ punti.get(j).distanzaP(punti.get(j+1));
	        }
	        somma=somma+getFine().distanzaP(punti.get(punti.size()-1));
	        return somma;
	    }
	}
}
